Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+            result.activeTips.erase(result.currentTip);
+          }
+        }
+        // If no more active tip, done with forward filtering; Otherwise, reset
+        // propagation state to track state at last tip of active tips
+        if (result.activeTips.empty()) {
+          ACTS_VERBOSE("Forward Kalman filtering finds "
+                       << result.trackTips.size() << " tracks");
+          result.forwardFiltered = true;
+        } else {
+          result.currentTip = result.activeTips.rbegin()->first;
+          ACTS_VERBOSE(
+              "Propagation jumps to branch with tip = " << result.currentTip);
+          reset(state, stepper, result);
+        }
+      }
+
+      // Post-processing after forward filtering
+      if (result.forwardFiltered) {
+        // Return error if forward filtering finds no tracks
+        if (result.trackTips.empty()) {
+          result.result =
+              Result<void>(CombinatorialKalmanFilterError::NoTracksFound);
+        } else {
+          if (not smoothing) {
+            // Manually set the targetReached to abort the propagation
+            ACTS_VERBOSE("Finish forward Kalman filtering");
+            state.navigation.targetReached = true;
+          } else {
+            // Iterate over the found tracks for smoothing and getting the
+            // fitted parameter. This needs to be accomplished in different
+            // propagation steps:
+            // -> first run smoothing for found track indexed with iSmoothed
+            if (not result.smoothed) {
+              result.currentTip = result.trackTips.at(result.iSmoothed);
+              ACTS_VERBOSE(
+                  "Finalize/run smoothing for track with entry index = "
+                  << result.currentTip);
+              // -> Sort the track states (as now the path length is set)
+              // -> Call the smoothing
+              // -> Set a stop condition when all track states have been handled
+              auto res = finalize(state, stepper, result);
+              if (!res.ok()) {
+                ACTS_ERROR("Error in finalize: " << res.error());
+                result.result = res.error();
+              }
+              result.smoothed = true;
+            }
+            // -> then progress to target/reference surface and built the final
+            // track parameters for found track indexed with iSmoothed
+            if (result.smoothed and
+                targetReached(state, stepper, *targetSurface)) {
+              ACTS_VERBOSE("Completing the track with entry index = "
+                           << result.currentTip);
+              // Transport & bind the parameter to the final surface
+              auto fittedState =
+                  stepper.boundState(state.stepping, *targetSurface, true);
+              // Assign the fitted parameters
+              result.fittedParameters.emplace(
+                  result.currentTip, std::get<BoundParameters>(fittedState));
```

Comment:
In principle, that's possible. We might add a track state with pure bound parameters at the end of each trajectory. Then I guess we need a pair of entry tips (for both the bound parameters and the real trajectory entry point) for each trajectory. But we might do this in another MR.

Reply:
Right, since we want both, and potentially differentiate between the two, the last 'real' track state and the reference track state. Ok let's leave it like this then.